README.md
# ibl-neuropixel
Collection of tools to handle Neuropixel 1.0 and 2.0 data
(documentation coming soon...)

## Installation
`pip install ibl-neuropixel`


## Destriping
### Getting started
This relies on a fast fourier transform external library: `pip install pyfftw`.

Minimal working example to destripe a neuropixel binary file.
```python
from pathlib import Path
from ibl.neurodsp.voltage import decompress_destripe_cbin
sr_file = Path('/datadisk/Data/spike_sorting/pykilosort_tests/imec_385_100s.ap.bin')
out_file = Path('/datadisk/scratch/imec_385_100s.ap.bin')

decompress_destripe_cbin(sr_file=sr_file, output_file=out_file, nprocesses=8)
```

### Viewer

The best way to look at the results is to use [viewephys](https://github.com/oliche/viewephys),
open an ephys viewer on the raw data.

- tick the destipe box.
- move to a desired location in the file
- ctr+P will make the gain and axis the same on both windows

![alt text](./docs/raw_bin_viewer_destripe.png "Ephys viewer")

You can then move within the raw data file.

### White Paper
The following describes the methods implemented in this repository.
https://doi.org/10.6084/m9.figshare.19705522

## Contribution

Pypi Release checklist:
```shell
flake8
rm -fR dist
rm -fR build
python setup.py sdist bdist_wheel
twine upload dist/*
#twine upload --repository-url https://test.pypi.org/legacy/ dist/*
```


release_notes.md
# 0.2.0

## 0.3.0
- Convert to namespace packaging.

## 0.2.2
- BUGFIX change getattr of `neuropixel` to allow for stared imports

## 0.2.1
- BUGFIX constant SYNC_PIN_OUT re-introduced in `ibl-neuropixel`

## 0.2.0
- destripe_decompress_cbin: add a parameter to output QC in a different location to accomodate pykilosort scratch dir
- support for NP2.4 and NP2.1 geometries on spikeGlx

## minor changes
-   deprecated SITES_COORDINATES that would default on Neuropixel 1.0 early generations (3A probes with 374 channels)
-   spikeglx compress: pass dtype to mtscomp


ibl-neurodsp/tests/unit/cpu/test_dsp.py
import unittest
import numpy as np
import scipy.signal
import scipy.fft

import ibl.neurodsp.fourier as fourier
import ibl.neurodsp.utils as utils
import ibl.neurodsp.voltage as voltage
import ibl.neurodsp.cadzow as cadzow
import ibl.neurodsp.smooth as smooth


class TestSyncTimestamps(unittest.TestCase):

    def test_timestamps_lin(self):
        np.random.seed(4132)
        n = 50
        drift = 17.14
        offset = 34.323
        tsa = np.cumsum(np.random.random(n) * 10)
        tsb = tsa * (1 + drift / 1e6) + offset

        # test linear drift
        _fcn, _drift = utils.sync_timestamps(tsa, tsb)
        assert np.all(np.isclose(_fcn(tsa), tsb))
        assert np.isclose(drift, _drift)

        # test missing indices on a
        imiss = np.setxor1d(np.arange(n), [1, 2, 34, 35])
        _fcn, _drift, _ia, _ib = utils.sync_timestamps(tsa[imiss], tsb, return_indices=True)
        assert np.all(np.isclose(_fcn(tsa[imiss[_ia]]), tsb[_ib]))

        # test missing indices on b
        _fcn, _drift, _ia, _ib = utils.sync_timestamps(tsa, tsb[imiss], return_indices=True)
        assert np.all(np.isclose(_fcn(tsa[_ia]), tsb[imiss[_ib]]))

        # test missing indices on both
        imiss2 = np.setxor1d(np.arange(n), [14, 17])
        _fcn, _drift, _ia, _ib = utils.sync_timestamps(tsa[imiss], tsb[imiss2], return_indices=True)
        assert np.all(np.isclose(_fcn(tsa[imiss[_ia]]), tsb[imiss2[_ib]]))


class TestParabolicMax(unittest.TestCase):
    # expected values
    maxi = np.array([np.NaN, 0, 3.04166667, 3.04166667, 5, 5])
    ipeak = np.array([np.NaN, 0, 5.166667, 2.166667, 0, 7])
    # input
    x = np.array([
        [0, 0, 0, 0, 0, np.NaN, 0, 0],  # some NaNs
        [0, 0, 0, 0, 0, 0, 0, 0],  # all flat
        [0, 0, 0, 0, 1, 3, 2, 0],
        [0, 1, 3, 2, 0, 0, 0, 0],
        [5, 1, 3, 2, 0, 0, 0, 0],  # test first sample
        [0, 1, 3, 2, 0, 0, 0, 5],  # test last sample
    ])

    def test_error_cases(self):
        pass

    def test_2d(self):
        ipeak_, maxi_ = utils.parabolic_max(self.x)
        self.assertTrue(np.all(np.isclose(self.maxi, maxi_, equal_nan=True)))
        self.assertTrue(np.all(np.isclose(self.ipeak, ipeak_, equal_nan=True)))

    def test_1d(self):
        # look over the 2D array as 1D chunks
        for i, x in enumerate(self.x):
            ipeak_, maxi_ = utils.parabolic_max(x)
            self.assertTrue(np.all(np.isclose(self.ipeak[i], ipeak_, equal_nan=True)))
            self.assertTrue(np.all(np.isclose(self.maxi[i], maxi_, equal_nan=True)))


class TestDspMisc(unittest.TestCase):

    def test_dsp_cosine_func(self):
        x = np.linspace(0, 40)
        fcn = utils.fcn_cosine(bounds=[20, 30])
        y = fcn(x)
        self.assertTrue(y[0] == 0 and y[-1] == 1 and np.all(np.diff(y) >= 0))


class TestPhaseRegression(unittest.TestCase):

    def test_fit_phase1d(self):
        w = np.zeros(500)
        w[1] = 1
        self.assertTrue(np.isclose(fourier.fit_phase(w, .002), .002))

    def test_fit_phase2d(self):
        w = np.zeros((500, 2))
        w[1, 0], w[2, 1] = (1, 1)
        self.assertTrue(np.all(np.isclose(fourier.fit_phase(w, .002, axis=0), np.array([.002, .004]))))
        self.assertTrue(np.all(np.isclose(fourier.fit_phase(w.transpose(), .002), np.array([.002, .004]))))


class TestShift(unittest.TestCase):

    def test_shift_already_fft(self):
        for ns in [500, 501]:
            w = scipy.signal.ricker(ns, 10)
            W = scipy.fft.rfft(w)
            ws = np.real(scipy.fft.irfft(fourier.fshift(W, 1, ns=np.shape(w)[0]), n=ns))
            self.assertTrue(np.all(np.isclose(ws, np.roll(w, 1))))

    def test_shift_floats(self):
        ns = 500
        w = scipy.signal.ricker(ns, 10)
        w_ = fourier.fshift(w.astype(np.float32), 1)
        assert w_.dtype == np.float32

    def test_shift_1d(self):
        ns = 500
        w = scipy.signal.ricker(ns, 10)
        self.assertTrue(np.all(np.isclose(fourier.fshift(w, 1), np.roll(w, 1))))

    def test_shift_2d(self):
        ns = 500
        w = scipy.signal.ricker(ns, 10)
        w = np.tile(w, (100, 1)).transpose()
        self.assertTrue(np.all(np.isclose(fourier.fshift(w, 1, axis=0), np.roll(w, 1, axis=0))))
        self.assertTrue(np.all(np.isclose(fourier.fshift(w, 1, axis=1), np.roll(w, 1, axis=1))))
        # # test with individual shifts for each trace
        self.assertTrue(
            np.all(np.isclose(fourier.fshift(w, np.ones(w.shape[1]), axis=0), np.roll(w, 1, axis=0))))
        self.assertTrue(
            np.all(np.isclose(fourier.fshift(w, np.ones(w.shape[0]), axis=1), np.roll(w, 1, axis=1))))


class TestSmooth(unittest.TestCase):

    def test_smooth_lp(self):
        np.random.seed(458)
        a = np.random.rand(500,)
        a_ = smooth.lp(a, [0.1, 0.15])
        res = fourier.hp(np.pad(a_, 100, mode='edge'), 1, [0.1, 0.15])[100:-100]
        self.assertTrue((utils.rms(a) / utils.rms(res)) > 500)


class TestFFT(unittest.TestCase):

    def test_spectral_convolution(self):
        sig = np.random.randn(20, 500)
        w = np.hanning(25)
        c = fourier.convolve(sig, w)
        s = np.convolve(sig[0, :], w)
        self.assertTrue(np.all(np.isclose(s, c[0, :-1])))

        c = fourier.convolve(sig, w, mode='same')
        s = np.convolve(sig[0, :], w, mode='same')
        self.assertTrue(np.all(np.isclose(c[0, :], s)))

        c = fourier.convolve(sig, w[:-1], mode='same')
        s = np.convolve(sig[0, :], w[:-1], mode='same')
        self.assertTrue(np.all(np.isclose(c[0, :], s)))

    def test_nech_optim(self):
        self.assertTrue(fourier.ns_optim_fft(2048) == 2048)
        self.assertTrue(fourier.ns_optim_fft(65532) == 65536)

    def test_freduce(self):
        # test with 1D arrays
        fs = np.fft.fftfreq(5)
        self.assertTrue(np.all(fourier.freduce(fs) == fs[:-2]))
        fs = np.fft.fftfreq(6)
        self.assertTrue(np.all(fourier.freduce(fs) == fs[:-2]))

        # test 2D arrays along both dimensions
        fs = np.tile(fourier.fscale(500, 0.001), (4, 1))
        self.assertTrue(fourier.freduce(fs).shape == (4, 251))
        self.assertTrue(fourier.freduce(np.transpose(fs), axis=0).shape == (251, 4))

    def test_fexpand(self):
        # test odd input
        res = np.random.rand(11)
        X = fourier.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(fourier.fexpand(X, 11)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test even input
        res = np.random.rand(12)
        X = fourier.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(fourier.fexpand(X, 12)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test with a 2 dimensional input along last dimension
        res = np.random.rand(2, 12)
        X = fourier.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(fourier.fexpand(X, 12)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test with a 3 dimensional input along last dimension
        res = np.random.rand(3, 5, 12)
        X = fourier.freduce(np.fft.fft(res))
        R = np.real(np.fft.ifft(fourier.fexpand(X, 12)))
        self.assertTrue(np.all((res - R) < 1e-6))
        # test with 2 dimensional input along first dimension
        fs = np.transpose(np.tile(fourier.fscale(500, 0.001, one_sided=True), (4, 1)))
        self.assertTrue(fourier.fexpand(fs, 500, axis=0).shape == (500, 4))

    def test_fscale(self):
        # test for an even number of samples
        res = [0, 100, 200, 300, 400, 500, -400, -300, -200, -100],
        self.assertTrue(np.all(np.abs(fourier.fscale(10, 0.001) - res) < 1e-6))
        # test for an odd number of samples
        res = [0, 90.9090909090909, 181.818181818182, 272.727272727273, 363.636363636364,
               454.545454545455, -454.545454545455, -363.636363636364, -272.727272727273,
               -181.818181818182, -90.9090909090909],
        self.assertTrue(np.all(np.abs(fourier.fscale(11, 0.001) - res) < 1e-6))

    def test_filter_lp_hp(self):
        # test 1D time serie: subtracting lp filter removes DC
        ts1 = np.random.rand(500)
        out1 = fourier.lp(ts1, 1, [.1, .2])
        self.assertTrue(np.mean(ts1 - out1) < 0.001)
        # test 2D case along the last dimension
        ts = np.tile(ts1, (11, 1))
        out = fourier.lp(ts, 1, [.1, .2])
        self.assertTrue(np.allclose(out, out1))
        # test 2D case along the first dimension
        ts = np.tile(ts1[:, np.newaxis], (1, 11))
        out = fourier.lp(ts, 1, [.1, .2], axis=0)
        self.assertTrue(np.allclose(np.transpose(out), out1))
        # test 1D time serie: subtracting lp filter removes DC
        out2 = fourier.hp(ts1, 1, [.1, .2])
        self.assertTrue(np.allclose(out1, ts1 - out2))

    def test_dft(self):
        # test 1D complex
        x = np.array([1, 2 - 1j, -1j, -1 + 2j])
        X = fourier.dft(x)
        assert np.all(np.isclose(X, np.fft.fft(x)))
        # test 1D real
        x = np.random.randn(7)
        X = fourier.dft(x)
        assert np.all(np.isclose(X, np.fft.rfft(x)))
        # test along the 3 dimensions of a 3D array
        x = np.random.rand(10, 11, 12)
        for axis in np.arange(3):
            X_ = np.fft.rfft(x, axis=axis)
            assert np.all(np.isclose(X_, fourier.dft(x, axis=axis)))
        # test 2D irregular grid
        _n0, _n1, nt = (10, 11, 30)
        x = np.random.rand(_n0 * _n1, nt)
        X_ = np.fft.fft(np.fft.fft(x.reshape(_n0, _n1, nt), axis=0), axis=1)
        r, c = [v.flatten() for v in np.meshgrid(np.arange(
            _n0) / _n0, np.arange(_n1) / _n1, indexing='ij')]
        nk, nl = (_n0, _n1)
        X = fourier.dft2(x, r, c, nk, nl)
        assert np.all(np.isclose(X, X_))


class TestWindowGenerator(unittest.TestCase):

    def test_window_simple(self):
        wg = utils.WindowGenerator(ns=500, nswin=100, overlap=50)
        sl = list(wg.firstlast)
        self.assertTrue(wg.nwin == len(sl) == 9)
        self.assertTrue(np.all(np.array([s[0] for s in sl]) == np.arange(0, wg.nwin) * 50))
        self.assertTrue(np.all(np.array([s[1] for s in sl]) == np.arange(0, wg.nwin) * 50 + 100))

        wg = utils.WindowGenerator(ns=500, nswin=100, overlap=10)
        sl = list(wg.firstlast)
        first = np.array([0, 90, 180, 270, 360, 450])
        last = np.array([100, 190, 280, 370, 460, 500])
        self.assertTrue(wg.nwin == len(sl) == 6)
        self.assertTrue(np.all(np.array([s[0] for s in sl]) == first))
        self.assertTrue(np.all(np.array([s[1] for s in sl]) == last))

    def test_nwindows_computation(self):
        for m in np.arange(0, 100):
            wg = utils.WindowGenerator(ns=500 + m, nswin=87 + m, overlap=11 + m)
            sl = list(wg.firstlast)
            self.assertTrue(wg.nwin == len(sl))

    def test_firstlast_slices(self):
        # test also the indexing versus direct slicing
        my_sig = np.random.rand(500,)
        wg = utils.WindowGenerator(ns=500, nswin=100, overlap=50)
        # 1) get the window by
        my_rms = np.zeros((wg.nwin,))
        for first, last in wg.firstlast:
            my_rms[wg.iw] = utils.rms(my_sig[first:last])
        # test with slice_array method
        my_rms_ = np.zeros((wg.nwin,))
        for wsig in wg.slice_array(my_sig):
            my_rms_[wg.iw] = utils.rms(wsig)
        self.assertTrue(np.all(my_rms_ == my_rms))
        # test with the slice output
        my_rms_ = np.zeros((wg.nwin,))
        for sl in wg.slice:
            my_rms_[wg.iw] = utils.rms(my_sig[sl])
        self.assertTrue(np.all(my_rms_ == my_rms))

    def test_tscale(self):
        wg = utils.WindowGenerator(ns=500, nswin=100, overlap=50)
        ts = wg.tscale(fs=1000)
        self.assertTrue(ts[0] == (100 - 1) / 2 / 1000)
        self.assertTrue((np.allclose(np.diff(ts), 0.05)))

    def test_rises_falls(self):
        # test 1D case with a long pulse and a dirac
        a = np.zeros(500,)
        a[80:120] = 1
        a[200] = 1
        # rising fronts
        self.assertTrue(all(utils.rises(a) == np.array([80, 200])))
        # falling fronts
        self.assertTrue(all(utils.falls(a) == np.array([120, 201])))
        # both
        ind, val = utils.fronts(a)
        self.assertTrue(all(ind == np.array([80, 120, 200, 201])))
        self.assertTrue(all(val == np.array([1, -1, 1, -1])))

        # test a 2D case with 2 long pulses and a dirac
        a = np.zeros((2, 500))
        a[0, 80:120] = 1
        a[0, 200] = 1
        a[1, 280:320] = 1
        a[1, 400] = 1
        # rising fronts
        self.assertTrue(np.all(utils.rises(a) == np.array([[0, 0, 1, 1], [80, 200, 280, 400]])))
        # falling fronts
        self.assertTrue(np.all(utils.falls(a) == np.array([[0, 0, 1, 1], [120, 201, 320, 401]])))
        # both
        ind, val = utils.fronts(a)
        self.assertTrue(all(ind[0] == np.array([0, 0, 0, 0, 1, 1, 1, 1])))
        self.assertTrue(all(ind[1] == np.array([80, 120, 200, 201, 280, 320, 400, 401])))
        self.assertTrue(all(val == np.array([1, -1, 1, -1, 1, -1, 1, -1])))


class TestVoltage(unittest.TestCase):

    def test_fk(self):
        """
        creates a couple of plane waves and separate them using the velocity HP filter
        """
        ntr, ns, sr, dx, v1, v2 = (500, 2000, 0.002, 5, 2000, 1000)
        data = np.zeros((ntr, ns), np.float32)
        data[:, :100] = scipy.signal.ricker(100, 4)
        offset = np.arange(ntr) * dx
        offset = np.abs(offset - np.mean(offset))
        data_v1 = fourier.fshift(data, offset / v1 / sr)
        data_v2 = fourier.fshift(data, offset / v2 / sr)

        noise = np.random.randn(ntr, ns) / 40
        fk = voltage.fk(data_v1 + data_v2 + noise, si=sr, dx=dx, vbounds=[1200, 1500],
                        ntr_pad=10, ntr_tap=15, lagc=.25)
        fknoise = voltage.fk(noise, si=sr, dx=dx, vbounds=[1200, 1500],
                             ntr_pad=10, ntr_tap=15, lagc=.25)
        # at least 90% of the traces should be below 50dB and 98% below 40 dB
        assert np.mean(20 * np.log10(utils.rms(fk - data_v1 - fknoise)) < -50) > .9
        assert np.mean(20 * np.log10(utils.rms(fk - data_v1 - fknoise)) < -40) > .98
        # test the K option
        kbands = np.sin(np.arange(ns) / ns * 8 * np.pi)
        kbands = np.sin(np.arange(ns) / ns * 8 * np.pi)
        fk = voltage.fk(data_v1 + data_v2 + noise + kbands, si=sr, dx=dx, vbounds=[1200, 1500],
                        ntr_pad=10, ntr_tap=15, lagc=.25,
                        kfilt={'bounds': [0, .01], 'btype': 'hp'})
        assert np.mean(20 * np.log10(utils.rms(fk - data_v1 - fknoise)) < -40) > .9

    def test_rcoeff(self):
        x = np.random.rand(2, 1000)
        y = x[0, :]
        r = np.corrcoef(x[1, :], y)
        assert voltage.rcoeff(x[0, :], y) == 1
        assert np.isclose(voltage.rcoeff(x[1, :], y), r[1, 0])
        assert np.all(np.isclose(voltage.rcoeff(x, y), r[0, :]))
        assert np.all(np.isclose(voltage.rcoeff(y, x), r[0, :]))


class TestCadzow(unittest.TestCase):

    def test_trajectory_matrixes(self):
        assert np.all(cadzow.traj_matrix_indices(4) == np.array([[1, 0], [2, 1], [3, 2]]))
        assert np.all(cadzow.traj_matrix_indices(3) == np.array([[1, 0], [2, 1]]))


class TestStack(unittest.TestCase):

    def test_simple_stack(self):
        ntr, ns = (24, 400)
        data = np.zeros((ntr, ns), dtype=np.float32)
        word = np.flipud(np.floor(np.arange(ntr) / 3))
        data += word[:, np.newaxis] * 10
        stack, fold = voltage.stack(data, word=word)
        assert(np.all(fold == 3))
        assert(np.all(np.squeeze(np.unique(stack, axis=-1)) == np.arange(8) * 10))
        # test with a header
        header = {'toto': np.random.randn(ntr)}
        stack, hstack = voltage.stack(data, word=word, header=header)
        assert(list(hstack.keys()) == ['toto', 'fold'])
        assert (np.all(hstack['fold'] == 3))


ibl-neurodsp/tests/unit/gpu/test_fourier.py
import numpy as np
import cupy as cp
import unittest

from ibl.neurodsp.fourier import fshift, channel_shift


class TestFourierAlignmentGpuCpuParity(unittest.TestCase):

    def test_parity(self):
        N_TIMES = 65600
        N_CHANNELS = 384

        data = np.random.randn(N_CHANNELS, N_TIMES) * 100
        sample_shifts = np.linspace(0, 0.9, N_CHANNELS)

        data_shifted_cpu = fshift(data, sample_shifts)
        data_shifted_gpu = channel_shift(cp.array(data), cp.array(sample_shifts))

        assert cp.allclose(data_shifted_gpu, data_shifted_cpu, atol=1e-5)


ibl-neurodsp/tests/unit/gpu/test_filter_gpu.py
import numpy as np
import cupy as cp
from scipy.signal import butter, sosfiltfilt
import unittest

from ibl.neurodsp.filter_gpu import sosfiltfilt_gpu


class TestFilterGpuCpuParity(unittest.TestCase):

    def test_parity(self):

        GPU_TOL = 1e-3
        BUTTER_KWARGS = {'N': 3, 'Wn': 300 / 30000 * 2, 'btype': 'highpass'}
        N_SIGNALS = 300
        N_SAMPLES = 60000

        sos = butter(**BUTTER_KWARGS, output='sos')
        array_cpu = np.cumsum(np.random.randn(N_SIGNALS, N_SAMPLES), axis=1)
        array_gpu = cp.array(array_cpu, dtype='float32')

        output_cpu = sosfiltfilt(sos, array_cpu)
        output_gpu = sosfiltfilt_gpu(sos, array_gpu)

        assert cp.allclose(output_cpu, output_gpu, atol=GPU_TOL)


ibl-neurodsp/ibl/neurodsp/smooth.py
import numpy as np
from scipy.interpolate import interp1d

import ibl.neurodsp.fourier as ft


def lp(ts, fac, pad=0.2):
    """
    Smooth the data in frequency domain (assumes a uniform sampling rate), using edge padding

    ibllib.dsp.smooth.lp(ts, [.1, .15])
    :param ts: input signal to be smoothed
    :param fac: 2 element vector of the frequency edges relative to Nyquist: [0.15, 0.2] keeps
    everything up to 15% of the full band tapering down to 20%
    :param pad: padding on the edges of the time serie, between 0 and 1 (0.2 means 20% of the size)
    :return: smoothed time series
    """
    # keep at least two periods for the padding
    lpad = int(np.ceil(ts.shape[0] * pad))
    ts_ = np.pad(ts, lpad, mode='edge')
    ts_ = ft.lp(ts_, 1, np.array(fac) / 2)
    return ts_[lpad:-lpad]


def rolling_window(x, window_len=11, window='blackman'):
    """
    Smooth the data using a window with requested size.

    This method is based on the convolution of a scaled window with the signal.
    The signal is prepared by introducing reflected copies of the signal
    (with the window size) in both ends so that transient parts are minimized
    in the beginning and end part of the output signal.

    :param x: The input signal
    :type x: list or numpy.array
    :param window_len: The dimension of the smoothing window,
                       should be an **odd** integer, defaults to 11
    :type window_len: int, optional
    :param window: The type of window from ['flat', 'hanning', 'hamming',
                   'bartlett', 'blackman']
                   flat window will produce a moving average smoothing,
                   defaults to 'blackman'
    :type window: str, optional
    :raises ValueError: Smooth only accepts 1 dimension arrays.
    :raises ValueError: Input vector needs to be bigger than window size.
    :raises ValueError: Window is not one of 'flat', 'hanning', 'hamming',
                        'bartlett', 'blackman'
    :return: Smoothed array
    :rtype: numpy.array
    """
    # **NOTE:** length(output) != length(input), to correct this:
    # return y[(window_len/2-1):-(window_len/2)] instead of just y.
    if isinstance(x, list):
        x = np.array(x)

    if x.ndim != 1:
        raise ValueError("smooth only accepts 1 dimension arrays.")

    if x.size < window_len:
        raise ValueError("Input vector needs to be bigger than window size.")

    if window_len < 3:
        return x

    if window not in ['flat', 'hanning', 'hamming', 'bartlett', 'blackman']:
        raise ValueError("Window is not one of 'flat', 'hanning', 'hamming',\
'bartlett', 'blackman'")

    s = np.r_[x[window_len - 1:0:-1], x, x[-1:-window_len:-1]]
    # print(len(s))
    if window == 'flat':  # moving average
        w = np.ones(window_len, 'd')
    else:
        w = eval('np.' + window + '(window_len)')

    y = np.convolve(w / w.sum(), s, mode='valid')
    return y[round((window_len / 2 - 1)):round(-(window_len / 2))]


def non_uniform_savgol(x, y, window, polynom):
    """Applies a Savitzky-Golay filter to y with non-uniform spacing as defined in x.
    This is based on
    https://dsp.stackexchange.com/questions/1676/savitzky-golay-smoothing-filter-for-not-equally-spaced-data
    The borders are interpolated like scipy.signal.savgol_filter would do
    https://dsp.stackexchange.com/a/64313
    Parameters
    ----------
    x : array_like
        List of floats representing the x values of the data
    y : array_like
        List of floats representing the y values. Must have same length as x
    window : int (odd)
        Window length of datapoints. Must be odd and smaller than x
    polynom : int
        The order of polynom used. Must be smaller than the window size
    Returns
    -------
    np.array
        The smoothed y values
    """

    if len(x) != len(y):
        raise ValueError('"x" and "y" must be of the same size')
    if len(x) < window:
        raise ValueError('The data size must be larger than the window size')
    if type(window) is not int:
        raise TypeError('"window" must be an integer')
    if window % 2 == 0:
        raise ValueError('The "window" must be an odd integer')
    if type(polynom) is not int:
        raise TypeError('"polynom" must be an integer')
    if polynom >= window:
        raise ValueError('"polynom" must be less than "window"')

    half_window = window // 2
    polynom += 1

    # Initialize variables
    A = np.empty((window, polynom))  # Matrix
    tA = np.empty((polynom, window))  # Transposed matrix
    t = np.empty(window)  # Local x variables
    y_smoothed = np.full(len(y), np.nan)

    # Start smoothing
    for i in range(half_window, len(x) - half_window, 1):
        # Center a window of x values on x[i]
        for j in range(0, window, 1):
            t[j] = x[i + j - half_window] - x[i]

        # Create the initial matrix A and its transposed form tA
        for j in range(0, window, 1):
            r = 1.0
            for k in range(0, polynom, 1):
                A[j, k] = r
                tA[k, j] = r
                r *= t[j]

        # Multiply the two matrices
        tAA = np.matmul(tA, A)
        # Invert the product of the matrices
        tAA = np.linalg.inv(tAA)
        # Calculate the pseudoinverse of the design matrix
        coeffs = np.matmul(tAA, tA)
        # Calculate c0 which is also the y value for y[i]
        y_smoothed[i] = 0
        for j in range(0, window, 1):
            y_smoothed[i] += coeffs[0, j] * y[i + j - half_window]

        # If at the end or beginning, store all coefficients for the polynom
        if i == half_window:
            first_coeffs = np.zeros(polynom)
            for j in range(0, window, 1):
                for k in range(polynom):
                    first_coeffs[k] += coeffs[k, j] * y[j]
        elif i == len(x) - half_window - 1:
            last_coeffs = np.zeros(polynom)
            for j in range(0, window, 1):
                for k in range(polynom):
                    last_coeffs[k] += coeffs[k, j] * y[len(y) - window + j]

    # Interpolate the result at the left border
    for i in range(0, half_window, 1):
        y_smoothed[i] = 0
        x_i = 1
        for j in range(0, polynom, 1):
            y_smoothed[i] += first_coeffs[j] * x_i
            x_i *= x[i] - x[half_window]

    # Interpolate the result at the right border
    for i in range(len(x) - half_window, len(x), 1):
        y_smoothed[i] = 0
        x_i = 1
        for j in range(0, polynom, 1):
            y_smoothed[i] += last_coeffs[j] * x_i
            x_i *= x[i] - x[-half_window - 1]

    return y_smoothed


def smooth_interpolate_savgol(signal, window=31, order=3, interp_kind='cubic'):
    """Run savitzy-golay filter on signal, interpolate through nan points.

    Parameters
    ----------
    signal : np.ndarray
        original noisy signal of shape (t,), may contain nans
    window : int
        window of polynomial fit for savitzy-golay filter
    order : int
        order of polynomial for savitzy-golay filter
    interp_kind : str
        type of interpolation for nans, e.g. 'linear', 'quadratic', 'cubic'
    Returns
    -------
    np.array
        smoothed, interpolated signal for each time point, shape (t,)
    """

    signal_noisy_w_nans = np.copy(signal)
    timestamps = np.arange(signal_noisy_w_nans.shape[0])
    good_idxs = np.where(~np.isnan(signal_noisy_w_nans))[0]
    # perform savitzky-golay filtering on non-nan points
    signal_smooth_nonans = non_uniform_savgol(
        timestamps[good_idxs], signal_noisy_w_nans[good_idxs], window=window, polynom=order)
    signal_smooth_w_nans = np.copy(signal_noisy_w_nans)
    signal_smooth_w_nans[good_idxs] = signal_smooth_nonans
    # interpolate nan points
    interpolater = interp1d(
        timestamps[good_idxs], signal_smooth_nonans, kind=interp_kind, fill_value='extrapolate')
    signal = interpolater(timestamps)

    return signal
